### DISCUSSION TITLE
uCNC on Arduino Uno with CNC Shield V3

### DISCUSSION THREAD

**Discussion by timint** [Paciente8159/uCNC#390] [General]
Hi, I was looking for a maintained and better updated alternative to GRBL and I found uCNC. Big thank you to the author of this project and for all your efforts in maintaining it. 🙂

I compiled uCNC 1.7.0-beta with Arduino 2.1.0 IDE.
I set the board to `BOARD_UNO_SHIELD_V3` in `cnc_config.h`.
I commented the PWM spindle override in `board_uno_shield_v3.h` (as I use the board's Limit Z pins for the Z limit switch).
Booting it throws non recoverable hard limit warnings in my CNC software. I use to get around soft limit errors with a configuration wizard, but that's a no go with hard limit errors.

This makes me wonder if I got it compiled right by setting the board in cnc_config.h?

![image](https://github.com/Paciente8159/uCNC/assets/359192/43785d91-6f71-4d25-9604-9753c8bd17e6)

![image](https://github.com/Paciente8159/uCNC/assets/359192/c504f222-cd09-479d-8099-6b2f61a8eeed)

Here are the pins I currently use:

![image](https://github.com/Paciente8159/uCNC/assets/359192/10410fb0-43fd-46bf-bbfa-07ac35f36003)

**Comment by Paciente8159** [comment]
Hi there.
I think this is related to this [wiki topic](https://github.com/Paciente8159/uCNC/wiki/Basic-user-guide#%C2%B5cnc-wiring)
By default and as a safety feature all controls, limits and probe pin are wired to be triggered if not connected to ground.
This prevents many issues like unconnected switches, faulty wires, some low levels of EMF noise, etc...

For this reason µCNC expects you to use Normally Closed (NC) switches to close this pins to ground. If you are using Normally Open (NO) these pins will remain in the triggered state.

There are several ways you can go around this.
- If you don't a certain group of pins you can disable them completely [here](https://github.com/Paciente8159/uCNC/blob/ef5fbfa8e0d8e176b9bc340b86b7efeec4016f6c/uCNC/cnc_config.h#L455)
- You can disable their internal pull up but that will leave them in a floating state and susceptible to false triggers due to environmental noise [here](https://github.com/Paciente8159/uCNC/blob/ef5fbfa8e0d8e176b9bc340b86b7efeec4016f6c/uCNC/cnc_hal_config.h#L56)
- You can invert their internal login either in compilation time [here](https://github.com/Paciente8159/uCNC/blob/ef5fbfa8e0d8e176b9bc340b86b7efeec4016f6c/uCNC/cnc_config.h#LL109C17-L109C17) or by configure the logic mask in the [settings](https://github.com/Paciente8159/uCNC/wiki/Basic-user-guide#%C2%B5cnc-configurations)

The other alternative is to do your customization in the[ web config generator](https://paciente8159.github.io/uCNC-config-builder/) tool but the latter option (to invert the pins logic mask) is not present there.

**Reply by timint** [reply]
Okay, that sounds reasonable that safety comes first. 🙂I took for granted this firmware was interchangable with GRBL out of the box.

I got these limit switches that I think are both ways?
![image](https://github.com/Paciente8159/uCNC/assets/359192/981511c0-c502-4f4a-b438-b6ec9fa49f19)

**Reply by Paciente8159** [reply]
The answer is not straight forward.
The switch circuit implements it's own logic.

The best way for you to debug the problem is using the firmware by it self like you would do on Grbl.

Since you are using Arduino IDE open it, setup your board and port and go to Tools>Serial Monitor
In serial monitor send a status query command by sending a single `?`
You should get a reply in a format similar to this:

`<Idle|MPos:0.000,0.000,0.000|FS:0.000,0|Pn:XY|WCO:0.000,0.000,0.000>`

The `PN:` portion gives you the logic state of the control, limit and probe pins. In my case as you can see X and Y limits are active. You can read about it [here](https://github.com/Paciente8159/uCNC/wiki/Basic-user-guide#%C2%B5cnc-push-messages).

For example with your X limit switch installed send the `?`. If in the `PN:` X is present that means it is triggered.
Press the x switch with your finger and while holding it pressed send another `?`. 

Did the X disapeared?
- If Yes then you need to invert the logic mask on setting [$5](https://github.com/Paciente8159/uCNC/wiki/Basic-user-guide#5----limit-pins-invert-mask).
- If no change occured then you have some kind of problem either in the board or in your switch/connection.

**Comment by timint** [comment]
Okay I made some good progress today.
* Swapped the the limit switches to normally open.
* Short circuited Abort and Hold

That silenced the alarms and now it only says "Alarm: Probe". But now I am able to Jog, wohoo!

With the standard GRBL firmware I had problems with everything freezing after a couple of seconds and I had to pull the power. With your firmware I managed to successfully run my first geometry test. No crash so far 👍👍

My current use of the board looks like this:
![image](https://github.com/Paciente8159/uCNC/assets/359192/64604ca9-af3d-4b92-9641-980405ed662a)

Two new questions:

- How can I enable the Z- limit pins on my board for the Z- limit switch? I had commented out the PWM part in `boardmap_uno_shield_v3.h` and tried restoring it back to defaults.
- Is the Probe alarm just a friendly reminder? I have no probe (yet) and wonder if there is anything I need to do regarding the alarm?

**Comment by Paciente8159** [comment]
The Shield V3 boardmap uses the boardmap configuration for Grbl 1.1, and modifies it to mimic Grbl 0.8 mapping since this shield was designed for that version.

Do not comment the PWM settings since pin D11 is used for Z-limit on Grbl 0.8 and for PWM on Grbl 1.1.
If you take a closer look boardmap shield V3 actually is disabling the PWM and not setting it up.

Using the original Shield V3 boardmap Z limit will be in the Z- and Z+ terminals as it should be.

Also the probe pin is configured for PIN A5 which is not routed on Shield V3. The active Probe alarm will have no effect unless you try to probe.

**Reply by timint** [reply]
I did try reverting boardmap_uno_shield_v3.h back to original but nothing happens when I connect the limit switch to the Z- pins. Usually it lights up a led when when I trigger the switch or trigger an alarm.

I will try flashing it again with the original code in case it wasn't done properly. Thank you for moving me forward. 🙂

**Reply by Paciente8159** [reply]
Use the original code and recompile and upload again. Same test. Remove any connection to the z- and z+ pins.
Send an `?`. What does the status message outputs.
Then connect z- to ground. Again a `?` command.
Did the Pn status changed?

**Comment by Paciente8159** [comment]
Just another note that just caught my eye.
Are you shunting EN/GND pins together???
If yes...remove it. You should not do that. Those pins are just there in case you what to rewire them to external stepper drivers.
The EN pin is controlled by the board pin D8.

**Reply by timint** [reply]
I was told EN/GND was necessary to enable the board. I recall somone also demonstrated that in a youtube video, and I have a memory of it not communicating if the pin is removed. But I trust you know this stuff a lot better. I will absolutely try to remove that tomorrow. Thanks for providing feedback.

**Reply by Paciente8159** [reply]
I don't know if that was the case with some very early version of Grbl or not but both Grbl 1.1 and uCNC control the steppers enable pins.
Shunting an output pin like the EN pin to ground can lead to bad things happening including damaging the chip.

**Reply by timint** [reply]
I have removed the EN/GND pin connector and the board is fully operational without it. 👍

**Comment by timint** [comment]
I have flashed the firmware with 1.7.1 and I also spotted the problem. I had defaulted back to BOARD_UNO.
With BOARD_UNO_SHIELD_V3 and the default settings I can confirm the Z- limit switch is fully operational.

Looking back at my journey I can see that you have kind of thought of everything. I wish I had begun with uCNC and not go through GRBL 0.8 and 1.1 as that just overcomplicated things for me. This project is a much better way to get started right out of the box. 🙂 I hope the community will recognize your project as a really awesome alternative to GRBL 1.1.

**Reply by Paciente8159** [reply]
Thanks for your feedback and support. If you find any problem while using it please report and I'll try to help the best way that I can.

**Comment by trihwangyudi** [comment]
i try install ucnc in arduino uno with ide 2, but i have error like this (i use linux mint una)

Sketch uses 33824 bytes (104%) of program storage space. Maximum is 32256 bytes.
Sketch too big; see https://support.arduino.cc/hc/en-us/articles/360013825179 for tips on reducing it.
Global variables use 1426 bytes (69%) of dynamic memory, leaving 622 bytes for local variables. Maximum is 2048 bytes.
text section exceeds available space in board

Compilation error: text section exceeds available space in board


try old ide same result 

Arduino: 1.8.7 (Linux), Board: "Arduino Uno"

Sketch uses 33824 bytes (104%) of program storage space. Maximum is 32256 bytes.
Global variables use 1426 bytes (69%) of dynamic memory, leaving 622 bytes for local variables. Maximum is 2048 bytes.

Sketch too big; see http://www.arduino.cc/en/Guide/Troubleshooting#size for tips on reducing it.
Error compiling for board Arduino Uno.

This report would have more information with
"Show verbose output during compilation"
option enabled in File -> Preferences.

this my file

[uCNC test.zip](https://github.com/Paciente8159/uCNC/files/13761135/uCNC.test.zip)

**Reply by Paciente8159** [reply]
Compiling µCNC for UNO with Arduino IDE will be very hard.
You will probably only be able to do it, disabling some features (2 axis version only or disabling control and/or limit switch pins), or with version 1.7.x of µCNC.
The flash size of the UNO is very limited.
Another alternative is compiling using VSCode. That uses some extra compiler flags that can achieve a bit more compression on the final binary.

**Reply by trihwangyudi** [reply]
i sucsessfully installed version 1.7.6 but i can't dual axis Y, i connect A step with collant enable, and A dir with sda pin, 
this my set up for board uno 
/*
	Name: boardmap_uno.h
	Description: Contains all MCU and PIN definitions for Arduino UNO similar to Grbl 1.1+ to run µCNC.

	Copyright: Copyright (c) João Martins
	Author: João Martins
	Date: 04/02/2020

	µCNC is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version. Please see <http://www.gnu.org/licenses/>

	µCNC is distributed WITHOUT ANY WARRANTY;
	Also without the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
	See the	GNU General Public License for more details.
*/

#ifndef BOARDMAP_UNO_H
#define BOARDMAP_UNO_H

#ifdef __cplusplus
extern "C"
{
#endif

#ifndef BOARD_NAME
#define BOARD_NAME "Arduino UNO"
#endif

// reduces RAM usage a bit to prevent hardware resets
#define PLANNER_BUFFER_SIZE 14

#define PCINT0_PORT B
#define PCINT1_PORT C
#define PCINT2_PORT D

// SAME AS GRBL for test purposes
// Setup step pins
#define STEP2_BIT 4	 // assigns STEP2 pin
#define STEP2_PORT D // assigns STEP2 port
#define STEP1_BIT 3	 // assigns STEP1 pin
#define STEP1_PORT D // assigns STEP1 port
#define STEP0_BIT 2	 // assigns STEP0 pin
#define STEP0_PORT D // assigns STEP0 port
#define STEP6_BIT 4 //assigns STEP6 pin (will mirror DUAL_AXIS0)
#define STEP6_PORT C //assigns STEP6 port (will mirror DUAL_AXIS0)

// Setup dir pins
#define DIR2_BIT 7	// assigns DIR2 pin
#define DIR2_PORT D // assigns DIR2 port
#define DIR1_BIT 6	// assigns DIR1 pin
#define DIR1_PORT D // assigns DIR1 port
#define DIR0_BIT 5	// assigns DIR0 pin
#define DIR0_PORT D // assigns DIR0 port

// Setup limit pins
#define LIMIT_Z_BIT 4  // assigns LIMIT_Z pin
#define LIMIT_Z_PORT B // assigns LIMIT_Z port
#define LIMIT_Z_ISR 0  // assigns LIMIT_Z ISR
#define LIMIT_Y2_BIT 4 //Z and second Y limit share the pin
#define LIMIT_Y2_PORT B //Z and second Y limit share the pin
#define LIMIT_Y2_ISR 0 //Z and second Y limit share the pin

#define LIMIT_Y_BIT 2  // assigns LIMIT_Y pin
#define LIMIT_Y_PORT B // assigns LIMIT_Y port
#define LIMIT_Y_ISR 0  // assigns LIMIT_Y ISR

#define LIMIT_X_BIT 1  // assigns LIMIT_X pin
#define LIMIT_X_PORT B // assigns LIMIT_X port
#define LIMIT_X_ISR 0  // assigns LIMIT_X ISR

// Setup probe pin
#define PROBE_BIT 5
#define PROBE_PORT C
#define PROBE_ISR 1

// Setup control input pins
// #define ESTOP_BIT 0
// #define FHOLD_BIT 1
// #define CS_RES_BIT 2
// #define ESTOP_PORT C
// #define FHOLD_PORT C
// #define CS_RES_PORT C
// #define ESTOP_ISR 1
// #define FHOLD_ISR 1
// #define CS_RES_ISR 1

// Setup com pins
#define RX_BIT 0
#define TX_BIT 1
#define RX_PORT D
#define TX_PORT D
#define RX_PULLUP
	// only uncomment this if other port other then 0 is used
	//#define UART_PORT 0

	// Setup PWM
#define PWM0_BIT 3	// assigns PWM0 pin
#define PWM0_PORT B // assigns PWM0 pin
#define PWM0_CHANNEL A
#define PWM0_TIMER 2

// Setup generic IO Pins
// Functionalities are set in cnc_hal_config.h file

// spindle dir
// #define DOUT0_BIT 5
// #define DOUT0_PORT B

// coolant
// #define DOUT2_BIT 3
// #define DOUT2_PORT C

	// // spindle speed sensor
	//  #define ANALOG0_BIT 4
	//  #define ANALOG0_PORT C
	//  #define ANALOG0_CHANNEL 4

// 	// servo type signal pin
// #define SERVO0_BIT 4
// #define SERVO0_PORT C

// // encoders
// #define DIN0_BIT 0
// #define DIN0_PORT C
// #define DIN0_ISR 1
// #define DIN8_BIT 1
// #define DIN8_PORT C

// Stepper enable pin. For Grbl on Uno board a single pin is used
#define STEP0_EN_BIT 0
#define STEP0_EN_PORT B

	// Setup the Step Timer used has the heartbeat for µCNC
	// Timer 1 is used by default
	//#define ITP_TIMER 1

	// Setup the RTC Timer used by µCNC to provide an (mostly) accurate time base for all time dependent functions
	// Timer 0 is set by default
	//#define RTC_TIMER 0

	#define ONESHOT_TIMER 2

#ifdef __cplusplus
}
#endif

#endif

after few hours i edited board map, but i cant compile its error like this :
Arduino: 1.8.7 (Linux), Board: "Arduino Uno"

In file included from sketch/src/cnc.h:144:0,
                 from /home/aguswatik/esp32shield/uCNC-1.7.6/uCNC/uCNC.ino:1:
sketch/src/cnc_hal_config_helper.h:669:2: error: #error "Enabling dual axis drive requires to configure at least one axis with dual drive"
 #error "Enabling dual axis drive requires to configure at least one axis with dual drive"
  ^~~~~
exit status 1
Error compiling for board Arduino Uno.

This report would have more information with
"Show verbose output during compilation"
option enabled in File -> Preferences.

**Comment by trihwangyudi** [comment]
[boardmap_uno.zip](https://github.com/Paciente8159/uCNC/files/13763995/boardmap_uno.zip)
this my board map

**Comment by Paciente8159** [comment]
Hi there.

First I recommend using the [1.7.x bugfix branch](https://github.com/Paciente8159/uCNC/tree/v1.7.x-bugfix)

To enable Y dual drive you need to enable it in cnc_hal_config.h file
In these lines [here](https://github.com/Paciente8159/uCNC/blob/5e5b3b4e5254b6d24454c8eda8125a0ef6c4cdb3/uCNC/cnc_hal_config.h#L100)

Modify it to this:

```
/**
 * Uncomment this feature to enable up to 2 dual drive axis
 * NOTE: If Laser PPI is enabled one of the stepper drivers position will be used by the laser controller
 * Usually that is STEPPER<AXIS_COUNT> so if AXIS_COUNT=3, STEPPER3 will be used by laser PPI
 */
#define ENABLE_DUAL_DRIVE_AXIS
#ifdef ENABLE_DUAL_DRIVE_AXIS
// defines the first dual drive capable axis
#define DUAL_DRIVE0_AXIS Y
// by default this will be rewired to STEPPER6 (if available on the board)
// this can be uncommented to re-wire to an available (unused stepper other then 6)
#define DUAL_DRIVE0_STEPPER 6
//  #define DUAL_DRIVE0_ENABLE_SELFSQUARING

// defines the first second drive capable axis
// #define DUAL_DRIVE1_AXIS Y
// by default this will be rewired to STEPPER7 (if available on the board)
// this can be uncommented to re-wire to an available (unused stepper other then 7)
// #define DUAL_DRIVE1_STEPPER 7
// #define DUAL_DRIVE1_ENABLE_SELFSQUARING
#endif
```

Your Y2 step output will be at pin A4 like you defined in your boardmap.

**Reply by trihwangyudi** [reply]
i have installed with arduino ide, but stepper just moving one direction, it can't moving forward, i had to read the confih hal file, but noting mention about direction, or i have share direction on y axes direction pin? i havent test it with share direction pin,

i have to say sorry if i have lot an dumb question...

**Reply by Paciente8159** [reply]
For the Y2 motor you can either share Y pin dir, or if you want to map Y2 dir to pin D13 of the Arduino UNO then you also need to add to your boardmap the STEP6 dir pin like this

```
// assign shield v3 A dir to pin D13(PB5)
#define DIR6_BIT 5
#define DIR6_PORT B
```

**Comment by trihwangyudi** [comment]
yes its work, and i have successfully compile.with makefile in linux,😄😄😄😄
another question is, i had a plan to change the main board, from arduino uno to wemos d1 r2, are those board have grbl touch or similary web controll?? i have joined discord maybe i will ask there

**Reply by Paciente8159** [reply]
> yes its work, and i have successfully compile.with makefile in linux,😄😄😄😄

Great!

> another question is, i had a plan to change the main board, from arduino uno to wemos d1 r2, are those board have grbl touch or similary web controll?? i have joined discord maybe i will ask there

Ok. There are [some limitations on the ESP8266](https://github.com/Paciente8159/uCNC/issues/550) due the nature of this MCU but I will try to guide in the best way possible. Yes you can join the discord and I can provide support from there and maybe improve performance for you use case.

About the web interface, currently there is a [pendant module](https://github.com/Paciente8159/uCNC-modules/tree/master/web_pendant) that is useful to use your mobile phone to jog/control the machine. It's possible to connect via WiFi to the board and use an external App like CNCjs or Candle to control the board and run Gcode.

**Comment by trihwangyudi** [comment]
This time to switch the board, i have a question about uno and wemos, my machine still use uno, with lead screw, 8 mm with 4 start, its came to slow... If i have push to 1.200 mm/second feedrate its loosing steps, so i have to down speed with 500 mm/second, if i switch to wemos or esp32 dev kit v1 are they can push speed for 1.200 mm/second, my machine information, x y z with 4 nema 23 3,6 Amps, tb6600 driver, power supply 24v 15amps, lead screw 8mm with 4 start 2 lead, acme nut run on bcnc linux mint

**Comment by Paciente8159** [comment]
> If i have push to 1.200 mm/second feedrate its loosing steps, so i have to down speed with 500 mm/second, if i switch to wemos or esp32 dev kit v1 are they can push speed for 1.200 mm/second, my machine information, x y z with 4 nema 23 3,6 Amps, tb6600 driver, power supply 24v 15amps, lead screw 8mm with 4 start 2 lead, acme nut run on bcnc linux mint

You need to make the math to check the limits. That is to relate the number of steps per mm (settings $10x - x for each axis) and the max speed in steps per min (settings $11x - x for each axis) (note that is mm/min not mm/s).

The Uno board should be able to get to 30K steps per second or near 40K steps per second in 16bit mode.

Lets say you are using a 1/16th microstepping in a 1,8º per step motor (200 steps per full rev) and no gear. With a rod with a pitch of 2mm that translates to a step/mm of 1600 (16 * 200 / 2 = 1600).

30000 steps/s @ 1600 steps/mm translates in to a max of 18,75 mm/s or 1125 mm/min.

The ESP have a higher stepping rate so they will be able to do faster step rates.
One detail about the ESP (both ESP8266 and ESP32). There is a PR #584 that changes the way steps are calculated and generated. it's still under tests but should be merged soon.
